// react-liquidchart/dist/Helpers/props.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.gradient = exports.fillAndStroke = exports.dShape = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var dShape = exports.dShape = _propTypes2.default.shape({
  dx: _propTypes2.default.number,
  dy: _propTypes2.default.number,
  fill: _propTypes2.default.string,
  stroke: _propTypes2.default.string,
  fontSize: _propTypes2.default.number
});

var fillAndStroke = exports.fillAndStroke = _propTypes2.default.shape({
  fill: _propTypes2.default.string,
  stroke: _propTypes2.default.string
});

var gradient = exports.gradient = _propTypes2.default.shape({
  type: _propTypes2.default.number,
  r: _propTypes2.default.number,
  x: _propTypes2.default.number,
  x1: _propTypes2.default.number,
  x2: _propTypes2.default.number,
  y: _propTypes2.default.number,
  y1: _propTypes2.default.number,
  y2: _propTypes2.default.number,
  cx: _propTypes2.default.number,
  cy: _propTypes2.default.number,
  fx: _propTypes2.default.number,
  fy: _propTypes2.default.number,
  stops: _propTypes2.default.arrayOf(_propTypes2.default.node)
});
