// react-liquidchart/dist/Chart/LiquidChart.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Chart = require('react-offcharts-core/Components/Chart');

var _Chart2 = _interopRequireDefault(_Chart);

var _ReactIf = require('react-offcharts-core/Components/ReactIf');

var _ReactIf2 = _interopRequireDefault(_ReactIf);

var _Gradients = require('react-offcharts-core/Components/Defs/Gradients');

var _Gradients2 = _interopRequireDefault(_Gradients);

var _guid = require('react-offcharts-core/Utils/guid');

var _guid2 = _interopRequireDefault(_guid);

var _Clip = require('./Clip');

var _Clip2 = _interopRequireDefault(_Clip);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

var _Shape = require('./Shape');

var _Shape2 = _interopRequireDefault(_Shape);

var _Liquid = require('./Liquid');

var _Liquid2 = _interopRequireDefault(_Liquid);

var _props = require('../Helpers/props');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var gradientId = (0, _guid2.default)();

var LiquidChart = function LiquidChart(props) {
  return _react2.default.createElement(
    _Chart2.default,
    {
      width: props.width,
      height: props.height,
      responsive: props.responsive,
      value: props.value,
      clipId: (0, _guid2.default)(),
      gradientId: gradientId
    },
    _react2.default.createElement(
      _Liquid2.default,
      {
        outerBound: 0.95,
        innerBound: 0.85,
        amplitude: props.amplitude,
        frequency: props.frequency,
        waveScaleLimit: props.waveScaleLimit,
        animationWavesTime: props.animationWavesTime,
        animationTime: props.animationTime,
        animationEase: props.animationEase
      },
      _react2.default.createElement(
        _Shape2.default,
        {
          type: props.type,
          outerStyle: props.outerStyle,
          liquidStyle: props.liquidStyle,
          wetStyle: props.wetStyle,
          dryStyle: props.dryStyle
        },
        _react2.default.createElement(_Text2.default, {
          postfix: props.postfix,
          showDecimal: props.showDecimal,
          fontSizes: props.fontSizes,
          legend: props.legend
        })
      ),
      _react2.default.createElement(
        _ReactIf2.default,
        { condition: props.gradient.type, el: _react2.default.createElement('g', null) },
        _react2.default.createElement(_Gradients2.default, _extends({
          id: gradientId
        }, props.gradient, {
          fill: props.liquidStyle.fill
        }))
      ),
      _react2.default.createElement(_Clip2.default, null)
    )
  );
};

LiquidChart.propTypes = {
  width: _propTypes2.default.number,
  height: _propTypes2.default.number,
  responsive: _propTypes2.default.bool,
  value: _propTypes2.default.number,
  amplitude: _propTypes2.default.number,
  frequency: _propTypes2.default.number,
  waveScaleLimit: _propTypes2.default.bool,
  animationWavesTime: _propTypes2.default.number,
  animationEase: _propTypes2.default.string,
  animationTime: _propTypes2.default.number,
  type: _propTypes2.default.number,
  outerStyle: _props.fillAndStroke,
  liquidStyle: _props.fillAndStroke,
  wetStyle: _props.fillAndStroke,
  dryStyle: _props.fillAndStroke,
  // deliminator: PropTypes.string,
  postfix: _propTypes2.default.string,
  legend: _propTypes2.default.string,
  showDecimal: _propTypes2.default.bool,
  fontSizes: _propTypes2.default.shape({
    value: _propTypes2.default.number,
    decimal: _propTypes2.default.number,
    postfix: _propTypes2.default.number,
    legend: _propTypes2.default.number
  }),
  gradient: _propTypes2.default.shape({
    type: _propTypes2.default.number,
    r: _propTypes2.default.number,
    x: _propTypes2.default.number,
    x1: _propTypes2.default.number,
    x2: _propTypes2.default.number,
    y: _propTypes2.default.number,
    y1: _propTypes2.default.number,
    y2: _propTypes2.default.number,
    cx: _propTypes2.default.number,
    cy: _propTypes2.default.number,
    fx: _propTypes2.default.number,
    fy: _propTypes2.default.number,
    stops: _propTypes2.default.arrayOf(_propTypes2.default.node)
  })
};

LiquidChart.defaultProps = {
  outerBound: 0.9,
  innerBound: 0.85,
  margin: 0.005,
  frequency: 2,
  amplitude: 4,
  waveScaleLimit: true,
  outerStyle: { fill: 'rgb(23,139,202)' },
  liquidStyle: { fill: 'rgb(23, 139, 202)' },
  dryStyle: { fill: 'rgb(4, 86, 129)' },
  wetStyle: { fill: 'rgb(164, 219, 248)' },
  gradient: {
    x1: 0,
    x2: 0,
    y1: 100,
    y2: 0
  }
};

exports.default = LiquidChart;
